### PR TITLE
CI: Add PR title checker

### PR DESCRIPTION
**Pull request by sharnoff**
Loosely adapted the workflow from what we have in neondatabase/cloud, but with the format changed to match what we've previously done in this repo.

As currently implemented, PR titles would _generally_ be required to look like:
```
component: Short message
```
or, at its most flexible:
```
component1/subsystem,component2/subsystem: Short message
```

There's two escape hatches:

1. If the PR title starts with `Revert: "`
2. If you include <code>\<!-- affects all --\></code> in the PR body — in case your PR is more broad

---

**Some context:**

1. When looking at recent commits, I find it very useful to be able to quickly determine the scope of previous changes.
3. Historically, the existing PR title format was maintained as a "best effort" sort of situation, with some exceptions slipping through
4. Recently, there's been a few occasions where I've asked for a PR title to be changed to match the existing convention.

It's better to have something automated require this, so this PR adds that.

---

**Notes for review:** This PR is mostly for discussion! We can change the existing convention if we want :)

I do think that enforcing a common format is valuable regardless, even if it's not the particular implementation initially used in this PR.

### CODE REVIEW

**Comment by mikhail-sakhnov** [comment]
Broadly looks good and I really like the idea. What do you think if we extract shell script from yaml to a separate files under `scripts` directory? We could re use them locally, for example by using https://pre-commit.com/

**Comment by sharnoff** [comment]
Hm, looks like we'd need a `commit-msg` hook instead of a `pre-commit` hook. I'm not so sure about it tbh -- especially because it's ok to have very simple commits in the branch, as long as they get squashed :)

Would be good to discuss?

**Comment by mikhail-sakhnov** [comment]
I mostly used reference to pre commit as an example why to extract script to external file.

though if we look on your point: pre commit hooks are skippable and the tool I mentioned is optional. E.g even you have config for it in the repo, you still need explicitly turn it on.

File: .github/workflows/pr-format-verification.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+              exit 0
+          fi
+
+          # If the title looks like 'Revert "..."', then give it a pass.
+          if grep -q '^Revert "' <(echo "$title") >/dev/null ; then
+              exit 0
+          fi
+
+          # Check that title matches the regex:
+          TITLE_REGEX='^[a-z0-9-.]+(/[a-z0-9-.]+)*(,[a-z0-9-.]+(/[a-z0-9-.]+)*)*: .*$'
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply .github/workflows/pr-format-verification.yml]: Improved this in 123ed021d0634529216cb34f8ba4e2b5c1bc9cb9 -- wdyt?

[the remark this answers is not in the document]
Comment (Omrigan) [reply .github/workflows/pr-format-verification.yml]: Looks good!

**Review by Omrigan — APPROVED** [review]

**Comment by sharnoff** [comment]
Planning to merge as-is, but we should discuss a pre-commit hook or something :)

**Pull request merged**